### ISSUE TITLE
Implement the functionality to store the data in an sqllite database and load it back

### ISSUE THREAD

**Issue by AKuederle** [learn-and-code/MoneyPy#11]
This is a "main" issue. This means there will likely be various smaller issues related to this one. Please link them back here.

**Comment by elbivio** [comment]
I suppose here we should make the helper functions for database and tables creation and the main function objects for read and write the database records, but i suppose that the database structure and the choice of the method of handle this should be done collectively, pure python is enough or should we use an ORM like SQLAlchemy?

**Comment by AKuederle** [comment]
Ideally we should create some sort of data object. If we do that using helper functions and running our own serializiation or using an ORM would be up to you. But SQLAlchemy is very convenient in my experience. If I would do the project on my own, I would definitely use it.

**Comment by AKuederle** [comment]
I was thinking a little bit about it, and I think it might be cool to have some sort of lazyloading wrapper, which could just store a db query until the data is actually needed. Then the data is loaded in memory.
Maybe we could also do some "magic" were certain operations are performed on db level, as long as the data is not loaded and on pandas/python level if there are E.g. a simple `groupby add`.

This could look like that:
```python
>>> query = db.query(category='food')  # This does not load the data yet
>>> query.balance # Would calculate the sum of all transactions on a db level
20 €
>>> query.head(100) # This would actually load data from the db and store it as part of the query instance
# Print the first 100 entries
>>> query.sum() # Now that the data is loaded we can either run the calculation in Python or SQl. We can test which one is faster
```

NOTE: These are just ideas, which go a little bit beyond just creating a database. But maybe you can incoperate them into your discussion :)

**Comment by AKuederle** [comment]
Haha we will get there eventually!

But yes you are right. Let's focus on the basics.

Could you build a little SQLalchemy model (maybe just with 3 fields) and
explain to the other people on that issue how it works? Then you would have
a common starting ground

I will help you of course, if you need anything

**Comment by rakshu27** [comment]
Hi, Now only I am trying to figure out what ORM is and about SQLAlchemy. I have understood some basics. Shouldn't we arrive at a data model now to achieve this?

**Comment by AKuederle** [comment]
Yes we need a simple Datamodel to get started with that. We had a small discussion about this on discord the other day. And we came to the conclusion, that we should start with a simple model with 4-6 fields (we can expand later).

My proposal for the fields would be: Date, currency, amount, recipient, info/subject
And then we can add labels and categories in a second step

Btw. if you don't know about something we are talking about or are struggling to understand a concept, please ask :) This should not be about accepting that we just do things a certain way, but actually about understanding, why certain decisions are made, so that you can apply the things you learn in future projects

**Comment by elbivio** [comment]
Akunderle, thanks for the info, is a start, 
i've opened a chat with rakshu and SirHarisTheGuy on discord, let me know if you want to join, 
about the code do you want to see it here before or should we start to make pull requests?

**Comment by AKuederle** [comment]
If you want me to join, invite me, but I don't have to. Regarding the code: You have to decide how you wnat to go about developing. As you are all working on the same code, I would say you just create a single branch (+maybe small branches when people are testing things). The easiest way would be to create this branch as part of this repository (and not make a fork). This allows everybody to contribute and see what you are doing. Comminicate, to avoid git fuck-ups! Let people now if you added new commits to the branch, so that they can pull/rebase before they continue working :)

And if you make important decisions in your Discord chat, update either the issue or the pull request with a short summary.
Let me know if you need help with anything!

**Comment by AKuederle** [comment]
Btw. this worked on in #18 :) Have a look there to be updated!
